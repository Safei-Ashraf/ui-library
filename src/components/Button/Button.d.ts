import { ReactNode } from "react";

export interface ButtonProps {
	variant?: "filled" | "outline" | "text";
	color?: "primary" | "secondary" | "neutral";
	bgColor?: string;
	textColor?: string;
	borderColor?: string;
	size?: "lg" | "md" | "sm";
	text?: string;
	extraClasses?: string;
	children?: ReactNode;
	onClick?: (e: React.MouseEvent<HTMLButtonElement>) => void;
	icon?: boolean;
	iconDirection?: "right" | "left";
	iconOnly?: boolean;
	isDisabled?: boolean;
}
