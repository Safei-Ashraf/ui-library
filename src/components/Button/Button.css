.btn {
	@apply rounded-md tracking-wide font-medium border-solid outline-transparent border-thick font-inter flex flex-row justify-center items-center cursor-pointer;
}
.reverse {
	@apply flex-row-reverse;
}
.lg {
	@apply px-lg py-md text-fs.20 h-4xl;
}
.md {
	@apply px-md py-xs text-fs.base h-xxl;
}
.sm {
	@apply px-sm py-xxs text-fs.14 h-8;
}
/*variant = filled */
.primary-filled {
	@apply text-white bg-primary border-primary-border hover:bg-primary-hover hover:border-primary-hover active:bg-primary-focus active:border-primary-focus;
}
.secondary-filled {
	@apply text-white bg-secondary  border-secondary-border hover:bg-secondary-hover hover:border-secondary-hover active:bg-secondary-focus active:border-secondary-focus;
}
.neutral-filled {
	@apply text-white bg-neutral border-neutral-light  hover:bg-neutral-hover hover:border-neutral-hover active:bg-neutral-focus active:border-neutral-focus;
}
.disabled-filled {
	@apply text-white bg-neutral-disabled-bg text-neutral-disabled-text border-transparent pointer-events-none;
}

/*variant = outline */

.primary-outline {
	@apply bg-transparent text-primary border-primary-border  hover:bg-primary-bg-light hover:text-primary-hover active:border-thick active:text-primary-focus active:border-primary-focus;
}
.secondary-outline {
	@apply bg-transparent text-secondary border-secondary-border hover:text-secondary-hover hover:bg-secondary-outline-hover hover:border-secondary-hover active:border-thick active:border-secondary-focus active:text-secondary-focus;
}
.neutral-outline {
	@apply bg-transparent text-neutral  border-neutral hover:bg-neutral-lighter hover:text-neutral-hover  active:border-thick;
}
.disabled-outline {
	@apply bg-transparent text-neutral border-neutral-lighter pointer-events-none opacity-0.3;
}

/*variant = text */

.primary-text {
	@apply bg-transparent text-primary border-transparent  hover:bg-primary-bg-light hover:text-primary-hover active:border-thick active:text-primary-focus;
}
.secondary-text {
	@apply bg-transparent text-secondary border-transparent hover:text-secondary-hover hover:bg-secondary-outline-hover  active:border-thick  active:text-secondary-focus;
}
.neutral-text {
	@apply bg-transparent text-neutral  border-transparent hover:bg-neutral-lighter hover:text-neutral-hover  active:border-thick;
}
.disabled-text {
	@apply bg-transparent text-neutral border-transparent pointer-events-none opacity-0.3;
}

