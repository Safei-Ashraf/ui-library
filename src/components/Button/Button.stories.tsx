import React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";

import { Button } from "./Button";

//default export that describes the component,
export default {
	title: "UI-Library/Button",
	component: Button,
	argTypes: {
		variant: {
			control: "inline-radio",
			options: ["filled", "outline", "text"],
		},
		size: {
			control: "inline-radio",
			options: ["lg", "md", "sm"],
		},
		color: {
			control: "inline-radio",
			options: ["primary", "secondary", "neutral"],
		},
		iconDirection: { control: "inline-radio", options: ["left", "right"] },
		isDisabled: { control: "boolean", type: "boolean" },
		text: { control: "text", type: "string" },
		icon: {
			table: {
				disable: true,
			},
		},
	},
} as ComponentMeta<typeof Button>;

const Template: ComponentStory<typeof Button> = (args) => <Button {...args} />;

// named exports that describe the stories.
export const ButtonMain = Template.bind({});
ButtonMain.args = {
	variant: "filled",
	color: "primary",
	size: "lg",
	text: "Click me",
	isDisabled: false,
};
export const ButtonWithIcon = Template.bind({});
ButtonWithIcon.args = {
	icon: true,
};
