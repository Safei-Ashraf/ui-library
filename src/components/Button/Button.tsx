import className from "classnames";
import StarIcon from "../StarIcon";
import { ButtonProps } from "./Button.d";
import "./Button.css";

export const Button = ({
	variant = "filled",
	size = "md",
	color = "primary",
	extraClasses,
	onClick,
	children,
	iconDirection = "right",
	iconOnly = false,
	isDisabled,
	text = "Click me",
	icon,
	...props
}: ButtonProps): JSX.Element => {
	const buttonClasses = className("btn", {
		"primary-filled": variant === "filled" && color === "primary",
		"secondary-filled": variant === "filled" && color === "secondary",
		"neutral-filled": variant === "filled" && color === "neutral",
		"disabled-filled": isDisabled && variant === "filled",
		"primary-outline": variant === "outline" && color === "primary",
		"secondary-outline": variant === "outline" && color === "secondary",
		"neutral-outline": variant === "outline" && color === "neutral",
		"disabled-outline": isDisabled && variant === "outline",
		"primary-text": variant === "text" && color === "primary",
		"secondary-text": variant === "text" && color === "secondary",
		"neutral-text": variant === "text" && color === "neutral",
		"disabled-text": isDisabled && variant === "text",
		lg: size === "lg",
		md: size === "md",
		sm: size === "sm",
		reverse: iconDirection === "right" && icon === true,
		extraClasses: extraClasses,
	});

	return (
		<button
			onClick={onClick}
			className={buttonClasses}
			disabled={isDisabled}
			{...props}
		>
			{icon && (
				<div className="icon-container">
					<StarIcon
						size={size}
						iconDirection={iconDirection}
						iconOnly={iconOnly}
					/>
				</div>
			)}
			{text && !iconOnly && <div>{text}</div>}
			{children}
		</button>
	);
};
