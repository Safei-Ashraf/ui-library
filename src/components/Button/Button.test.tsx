import renderer from "react-test-renderer";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom";
import Button from "./index";

describe("Button", () => {
	it("render button", () => {
		render(<Button />);
	});
	it("render button with custom text content", () => {
		const text = "Hello Button";
		render(<Button text={text} />);

		expect(screen.getByText(text)).toBeInTheDocument();
	});
	it("render disabled button", () => {
		render(<Button isDisabled />);

		expect(screen.getByRole("button")).toBeDisabled();
	});
	it("does not render icon by default", () => {
		render(<Button />);
		expect(screen.queryByRole("icon")).toBeNull();
	});
	it("renders icon when prop provided", () => {
		render(<Button icon />);
		expect(screen.getByRole("icon")).toBeInTheDocument();
	});
	it("renders large size button with primary color and variant filled", () => {
		render(<Button size="lg" variant="filled" color="primary" />);
		expect(screen.getByRole("button")).toHaveClass("primary-filled", "lg");
	});
	it("renders md size button with neutral color and variant outline", () => {
		render(<Button size="md" variant="outline" color="neutral" />);
		expect(screen.getByRole("button")).toHaveClass("neutral-outline", "md");
	});
	it("renders sm size button with secondary color and variant text", () => {
		render(<Button size="sm" variant="text" color="secondary" />);
		expect(screen.getByRole("button")).toHaveClass("secondary-text", "sm");
	});
	it("check onClick function is called when user click", () => {
		const handleclick = jest.fn();
		render(<Button onClick={handleclick} />);
		userEvent.click(screen.getByRole("button"));
		expect(handleclick).toBeCalledTimes(1);
	});
});

describe("renders correctly", () => {
	const tree = renderer.create(<Button />).toJSON();
	expect(tree).toMatchSnapshot();
});
