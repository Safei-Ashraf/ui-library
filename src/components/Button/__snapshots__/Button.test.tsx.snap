// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<button
  className="btn primary-filled md"
  style={
    Object {
      "backgroundColor": "",
      "borderColor": "",
      "color": "",
    }
  }
>
  <div>
    Click me
  </div>
</button>
`;
